File: main/094-20200110.rb
def inorder_traversal(root)
  result = []
  white, black = 0, 1
  stack = [[white, root]]

  while !stack.empty?
    color, node = stack.pop

    next if !node

    if color == white
      stack.push([white, node.right]) # if node.right
      stack.push([black, node]) # if node
      stack.push([white, node.left]) # if node.left
    else
      result.push(node.val)
    end
    # puts stack.inspect
  end

  result
end

File: main/094-20200110.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func inorderTraversal(root *TreeNode) []int {
	var result []int
	process(root, &result)
	return result
}

func process(root *TreeNode, result *[]int) {
	if root == nil {
		return
	}
	if root.Left != nil {
		process(root.Left, result)
	}

	(*result) = append((*result), root.Val)

	if root.Right != nil {
		process(root.Right, result)
	}
}
